rp2040-hal/src/gpio.rs

```diff
@@ -0,0 +1,132 @@
+//! General Purpose Input and Output (GPIO)
+//!
+//! TODO: Docs for this whole module
+//!
+//! For example, to turn on the LED on a Pico board:
+//!
+//! ```rust
+//! let mut pac = rp2040_pac::Peripherals::take().unwrap();
+//!
+//! let pins = pac.IO_BANK0.split(&mut pac.RESETS);
+//! let mut led_pin = pins.gpio25.into_output(&mut pac.PADS_BANK0, &mut pac.SIO);
+//!  led_pin.set_low().unwrap();
+//! ```
+use core::convert::Infallible;
+use core::marker::PhantomData;
+use embedded_hal::digital::v2::OutputPin;
+
+pub struct Floating;
+pub struct Input<MODE> {
+    _mode: PhantomData<MODE>,
+}
+pub struct Output;
+
+pub trait GpioExt {
+    type Parts;
+
+    // TODO: Do we need a marker to check that clocks are up?
+    fn split(self, reset: &mut rp2040_pac::RESETS) -> Self::Parts;
+}
+
+const FUNCTION_SIO: u8 = 5;
+
+macro_rules! gpio {
+    ($GPIOX:ident, $gpiox:ident, $PADSX:ident, [
+        $($PXi:ident: ($pxi:ident, $i:expr),)+
+    ]) => {
+        impl GpioExt for pac::$GPIOX {
+            type Parts = Parts;
+
+            fn split(self, resets: &mut pac::RESETS) -> Parts {
+                resets.reset.modify(|_, w| w.$gpiox().clear_bit());
+                Parts {
+                    $(
+                        $pxi: $PXi { _mode: PhantomData },
+                    )+
+                }
+            }
+        }
+
+        pub struct Parts {
+            $(
+                pub $pxi: $PXi<Input<Floating>>,
+            )+
+        }
+
+        $(
+            pub struct $PXi<MODE> {
+                _mode: PhantomData<MODE>,
+            }
+
+            impl<MODE> $PXi<MODE> {
+                pub fn into_output(
+                    self,
+                    pads: &mut pac::$PADSX,
```

Comment:
I think we could do that. `GpioExt`'s `split` could require moves of `PADS` and `SIO`. I'll give that a shot and see if it causes grief.

I think stealing ownership of `PADS` is fine. I was hesitant about stealing `SIO` because it has a few other non-GPIO features (spinlocks and FIFOs between the cores). Things are even a bit more confusing because there's really multiple `SIO`s, one for each core. But maybe we just leave that alone for now, steal `SIO` and revisit it later once multi-core stuff is feasible (seems a bit far off for the HAL right now).

Reply:
Oh, I see. I did not check the PAC before about SIO. So in the long run we would need to split up the SIO registers into its parts. Not sure if that should be done on the PAC or the HAL level.

But as you said we can deal with that later.